src/meepgeom.cpp

```diff
@@ -1191,6 +1255,251 @@ void geom_epsilon::eff_chi1inv_matrix(meep::component c, symm_matrix *chi1inv_ma
+  int oi;
+  geom_box_tree tp;
+  fallback = false;
+
+  /* first let's check if our pixel has more than a single
+  interface (i.e. a corner) in which case we abandon any
+  smoothing efforts*/
+  if (!get_front_object(v, geometry_tree, p, &o, shiftby, mat, mat_behind, p_mat, p_mat_behind)) {
+  noavg:
+    get_material_pt(mat, v.center());
+  trivial:
+    material_epsmu(meep::type(c), mat, &meps, chi1inv_matrix);
+    material_gc(mat);
+    return;
+  }
+
+  /* we may have entered this routine (e.g. because of a material grid
+  backprop routine) but don't want to actually do any averaging */
+  if (maxeval == 0) {
+    get_material_pt(mat, v.center());
+    if (is_material_grid(mat)) {
+      tp = geom_tree_search(vec_to_vector3(v.center()), restricted_tree, &oi);
+      uval = matgrid_val(vec_to_vector3(v.center()), tp, oi, mat);
+    mgavg:
+      dual_interpolate_mat(mat, meps, uval);
+      sym_matrix_invert(chi1inv_matrix, &meps);
+      // get_material_pt(mat, v.center());
+      // material_epsmu(meep::type(c), mat, &meps, chi1inv_matrix);
+      material_gc(mat);
+      return;
+    }
+    else { goto trivial; }
+  }
+
+  // For variable materials with do_averaging == true, switch over to slow fallback integration
+  // method. For material grids, however, we have to do some more logic first...
+  if ((is_variable(mat) && mat->do_averaging) ||
+      (is_variable(mat_behind) && mat_behind->do_averaging)) {
+    if ((!is_material_grid(mat)) && (!is_material_grid(mat_behind))) {
+      fallback = true;
+      return;
+    }
+  }
+
+  /* ------------------------------------------- */
+  // One material
+  /* ------------------------------------------- */
+  /* check for case of only one object/material
+   in the case of the material grid, make sure we don't need
+   to do any interface averaging within.
+   */
+  if (material_type_equal(mat, mat_behind)) {
+    /* if we have a material grid, we need to calculate the fill fraction,
+    normal vector, etc. we also need to determine if we really need
+    to do any averaging, or if the current pixel is smooth enough
+    to just evaluate.
+    */
+    if (is_material_grid(mat)) {
+      // if we have a damping term and β≠∞, we can't smooth
+      if ((mat->damping != 0) && (mat->eta != std::numeric_limits<double>::infinity())) goto noavg;
+
+      tp = geom_tree_search(p_mat, restricted_tree, &oi);
+      uval = matgrid_val(p_mat, tp, oi, mat);
+      normal = matgrid_grad(p_mat, tp, oi, mat);
+      if (cvector3_re(normal).x == 0 && cvector3_re(normal).y == 0 && cvector3_re(normal).z == 0)
+        /* couldn't get normal vector for this point; no averaging is needed,
+        but we still need to interpolate onto our grid using duals */
+        goto mgavg;
+      duals::duald u_prime = normalize_dual(normal);
+      duals::duald d = (mat->eta - uval) / u_prime;
+      double r = v.diameter() / 2;
+      fill = get_material_grid_fill(v.dim, d, r, uval, mat->eta);
```

Comment:
Ok I think I have finally narrowed down the reason why some tests are failing.

Note that the `matgrid_grad` is especially sensitive to roundoff error (it calls several routines from both `meep` and `libctl`). There are many cases where the normal vector should be `(0,0,0)` (e.g. a uniform design region) but is typically off by a few $\epsilon$ in a random direction. **Especially when the design variables are uniform, which is the case for the failing tests**.

This is especially problematic when `uval`≈`mat->eta`, because then we get some weird $\frac{0}{0}$ conditions (this produces some dramatic swings in both the value for `d` and its dual derivative).

We need to think about some ways to make this part of the code more robust to roundoff error. This is a bit outside of my wheelhouse...

@stevengj, @oskooi, do you have any suggestions?

Reply:
As an example, suppose I'm optimizing a uniform design field (such that `rho=0.5` everywhere). Of course, I'm doing TO, so I run this through a conic filter first. Ideally, my output is also `rho_tilde=0.5` everywhere, but in reality there are some perturbations of the order of $\epsilon$ scattered throughout.

Now, because of this slight variation, when I try to compute the normal vector at a particular point `r` (i.e. the *spatial* gradient of the design field) using `matgrid_grad`, I *hope* to get a vector that looks like `(0,0,0)` but most of the time in these cases it looks something like `(3.330669e-15, 6.661338e-15, 0.000000e+00)`, where again, I'm off by a few $\epsilon$ in the `x` and `y` directions.

Furthermore, when I try to compute $d=\frac{\eta-\rho}{|\nabla\rho|}$, the quantity $\eta-\rho$ is *also* sensitive to floating point roundoff, and I occasionally get values like `-2.775558e-16`. Which means I have a really small number divided by an even smaller number ($|\nabla\rho|$ is pretty small), and we get some interesting behavior.

I've tried some thresholded compare statements (i.e. create a function `equals(double var1, double var2)` that checks if `var1` and `var2` are within a few $\epsilon$ of each other). Of course, these methods are a bit ad hoc and not ideal.

I'm thinking there's another way to tackle this problem that lets us avoid the issue of checking for roundoff altogether... just not sure what that is yet...

Reply:
Why don't we try adding a check to the end of [`meep::vec material_grid_grad`](https://github.com/NanoComp/meep/blob/ea43f83c2f8c3e74aab2f8dae3e2ed96b7347777/src/meepgeom.cpp#L366) to set the gradient vector to zero if all the components happen to be smaller than some threshold value? That way, we just need to modify only one place in the code. (For consistency, we also should do the same thing for other functions which compute the normal vector i.e. [`material_function::normal_vector`](https://github.com/NanoComp/meep/blob/ea43f83c2f8c3e74aab2f8dae3e2ed96b7347777/src/anisotropic_averaging.cpp#L44-L68)).

Reply:
I added a function, `AreEqual()`, which checks two numbers (I templated it so that it works with doubles, floats, etc.) within $2\epsilon$. This threshold (like all thresholds) is a bit arbitrary though...

This mitigates many of the errors I've seen so far. But there are still some issues slipping through the cracks, again due to roundoff error.

Reply:
In general, I would expect our final effective ε to be insensitive to errors like this, because in a region where ε is nearly constant its arithmetic and harmonic means are nearly equal so the choice of the normal vector becomes irrelevant.   So, we should have a well-conditioned function to differentiate.

This seems like a classic type of problem in numerical analysis where you have some nice well-conditioned function that you want to compute, but your computational algorithm involves some intermediate step (like the computation of `d`) that is badly conditioned, and screws up your calculation.  The solution is usually to reformulate the calculation algebraically to avoid the ill-conditioned step, e.g. to replace `d` with `d√|∇ρ|`.  Sometimes it requires some clever algebra, e.g. see the `exp1` and quadratic-formula examples in [these notes](https://nbviewer.org/github/mitmath/18335/blob/spring21/notes/Floating-Point-Intro.ipynb#Cancellation-error).